File: app/src/main/java/de/jerleo/database/TableReading.java
package de.jerleo.database;

import android.content.ContentValues;
import android.database.Cursor;

import java.util.ArrayList;
import java.util.List;

import de.jerleo.android.DateHelper;
import de.jerleo.database.DatabaseColumn.DataType;
import de.jerleo.model.Reading;

import static android.provider.BaseColumns._ID;
import static de.jerleo.database.Constants.COUNT;
import static de.jerleo.database.Constants.DATE;
import static de.jerleo.database.Constants.METER;
import static de.jerleo.database.Constants.READING;

public class TableReading extends DatabaseTable {

    public TableReading(Database database) {

        super(database, READING, true);

        addColumn(new DatabaseColumn(METER, DataType.INTEGER));
        addColumn(new DatabaseColumn(DATE, DataType.TEXT));
        addColumn(new DatabaseColumn(COUNT, DataType.INTEGER));

        final DatabaseColumn meter = getColumn(METER);
        meter.setForeignKeyConstraint(true);
        meter.setForeignTable(METER);
        meter.setForeignColumn(_ID);
    }

    @Override
    public ContentValues getValues(Object object, boolean forUpdate) {

        final Reading reading = (Reading) object;

        final ContentValues values = new ContentValues();
        if (forUpdate)
            values.put(_ID, reading.getId());
        values.put(METER, reading.getMeter().getId());
        values.put(DATE, DateHelper.getDatabaseDate(reading.getDate()));
        values.put(COUNT, reading.getCount());

        return values;
    }

    @Override
    public String getWhereClause(Object object) {

        final Reading reading = (Reading) object;
        return _ID + " = " + reading.getId();
    }

    @Override
    public List<Reading> read(String condition) {

        final ArrayList<Reading> readings = new ArrayList<>();
        final String[] columns = {_ID, DATE, COUNT};
        final String where = METER + " = " + condition;
        final Cursor cursor = database.query(READING, columns, where, null,
                null, null, null);
        while (cursor.moveToNext()) {
            final Reading reading = new Reading();
            int col = 0;
            reading.setId(cursor.getLong(col++));
            reading.setDate(DateHelper.getDatabaseDate(cursor.getString(col++)));
            reading.setCount(cursor.getInt(col));
            readings.add(reading);
        }
        return readings;
    }
}

File: app/src/main/java/de/jerleo/database/table/TableReading.kt
package de.jerleo.database.table

import android.content.ContentValues
import android.provider.BaseColumns
import de.jerleo.android.DateHelper
import de.jerleo.database.Column
import de.jerleo.database.Constants
import de.jerleo.database.Database
import de.jerleo.database.Table
import de.jerleo.model.Reading

class TableReading(database: Database) :
    Table(database, Constants.READING, true) {

    init {
        add(Column(Constants.METER, Column.Type.INTEGER).apply {
            isForeignKeyConstraint = true
            foreignTable = Constants.METER
            foreignColumn = BaseColumns._ID
        })
        add(Column(Constants.DATE, Column.Type.TEXT))
        add(Column(Constants.COUNT, Column.Type.INTEGER))
    }

    override fun values(obj: Any, forUpdate: Boolean): ContentValues {
        val reading = obj as Reading
        val values = ContentValues()
        if (forUpdate) values.put(BaseColumns._ID, reading.id)
        values.put(Constants.METER, reading.meter.id)
        values.put(Constants.DATE, DateHelper.format(reading.date))
        values.put(Constants.COUNT, reading.count)
        return values
    }

    override fun whereClause(obj: Any): String = BaseColumns._ID + " = " + (obj as Reading).id

    override fun read(condition: String?): ArrayList<Reading> {
        val readings = ArrayList<Reading>()
        val columns = arrayOf(BaseColumns._ID, Constants.DATE, Constants.COUNT)
        val where = Constants.METER + " = " + condition
        val cursor = database.query(
            Constants.READING, columns, where, null,
            null, null, null
        )
        while (cursor.moveToNext()) {
            val reading = Reading()
            var col = 0
            reading.id = cursor.getLong(col++)
            reading.date = DateHelper.parse(cursor.getString(col++))
            reading.count = cursor.getInt(col)
            readings.add(reading)
        }
        return readings
    }
}